### DISCUSSION TITLE
How to access the strategy of the trainer

### DISCUSSION THREAD

**Discussion by juliendenize** [Lightning-AI/pytorch-lightning#11272] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi, I am trying to make my code invariant to the choice of strategies by being able to compute the global batch size which depends on the strategy. For example, for DDP it is `N * batch_size` with N being the number of processes.

The use case I can think of is using the global batch size to initialize the optimizer.

```python
trainer(num_nodes=1, gpus=2, strategy='ddp') # pass the strategy ddp for example

class MyLightningModule(pl.LightningModule):
    @property
    def global_batch_size(self) -> int:
        if self.trainer.strategy is None:
            return self.trainer.datamodule.train.loader.batch_size
        elif self.trainer.strategy is DDPStrategy:
            return self.trainer.num_nodes * self.trainer.gpus *\       # There might be a better way to compute the
                      self.trainer.datamodule.train.loader.batch_size  # number of processes using the strategy
        ...

    def configure_optimizers(self) -> Dict[Any, Any]:
        optimizer, scheduler = hydra.utils.instantiate(
            self.hparams.optimizer, model=self, batch_size=self.global_batch_size, _recursive_=False)

        return {
            'optimizer': optimizer,
            'lr_scheduler': scheduler
        }
```

To do so, I would like to retrieve inside my Lightning module the strategy used by my trainer. I tried to find in the trainer code how to access the strategy and I found the property:

```python
# in pytorch_lightning.trainer.trainer.py
class Trainer(...):
    ...
    @property
    def strategy(self) -> Strategy:
        return self._accelerator_connector.strategy
```

However `self.trainer.strategy` in `configure_optimizers` raises `AttributeError: 'Trainer' object has no attribute 'strategy'`.

Weirdly, `self.trainer._accelerator_connector.strategy` works and returns the passed strategy in the trainer. Yet, if I understood correctly the `_accelerator_connector` should resolve the strategy `'ddp'` to `DDPStrategy` in its initialization but it returns `'ddp'`:
 
```python
# in pytorch_lightning.trainer.connectors.accelerator_connector.py

class AcceleratorConnector(...):
    def __init__(...):
        ...
        self.strategy = self.final_strategy()
     ...
```

Is it possible to access the strategy used for training?

**Comment by rohitgr7** [comment, marked as answer]
just out of curiosity, what sort of scheduler/optimizer are you initializing using the global_batch_size?

**Reply by juliendenize** [reply]
In my optimizer factory, I use the `global_batch_size` to scale the initial learning rate according to a specified rule of thumbs in my config such as linear scaling or square root scaling. 

I want my user to be free from computing the learning rate each time they want to change the batch size. For now, I pass in my config both the `local_batch_size` and the `global_batch_size` but I can't help to think that only one of these could be passed and the rest be agnostic for the user. However, due to the difference between `DDP` and `DP` or `DDP2` strategy, I didn't find how yet.

**Reply by rohitgr7** [reply]
okay!
`strategy` was something added recently and only available on master or will be available in the next big release.
Also it is available right after the trainer initialization.
```py
trainer = Trainer(accelerator='cpu', num_processes=4)
print(trainer.strategy)
```
output:
```
<pytorch_lightning.strategies.ddp_spawn.DDPSpawnStrategy at 0x7fc993d0e910>
```
so maybe can you try master? or if you are using an older version, maybe try `trainer.training_type_plugin` instead

**Reply by juliendenize** [reply]
Thank you it worked, I thought it was already fully released in version 1.5.7 and not only to specify the `training_type_plugin`. 

I don't think the `training_type_plugin` could be specified through the trainer API, so I guess it was resolved in the trainer init. As far as I understand, the strategy field allows more freedom for the user. Does this mean that in the next release `strategy` will replace `training_type_plugin`? 

I am new to pytorch_lightning so I don't quite understand the difference between the two, sorry if it is a naive question.

**Reply by rohitgr7** [reply]
yes, training_type_plugin is now strategy on master and will be available in the next major release.
1.5.x releases are patch fixes so we don't include any major update or feature/enhancement in that, only bug fixes.
